### PR TITLE
Extending multi-io support

### PR DESCRIPTION
**Pull request by elanmart**
Extends #1018 and #1035. 

The PR includes:
- Simple modification to `Graph` API (only one `input` argument, modify `set_previous` and allow more flexible connectivity using multi-io layers
- `MultiInputOutputLayer` (name very much [WIP]) 

The goal was to allow arbitrary connectivity between layers inside a `Graph`.
I think it is more intuitive than `join` mode inside a `Merge`.

Also, the new `MultiInputOutputLayer` has two additional advantages: it replicates `Graph` API and allows users to easily create new layers.

Below you can find a simple example of using the new API to build an `attention` module (will become usefull as soon as we figure out how to write a `Recurrent` container):

```python
class Attention(MultiInputOutoutLayer):
    """ Given a hidden state to condition on and sequence to attend, returns a context C 
    beign a weighted average of the attended"""
    
    def __init__(self, hidden_dim, init='glorot_uniform', activation='tanh', weights=None,
                 input_dim=None, **kwargs):
        self.init = initializations.get(init)
        self.activation = activations.get(activation)
        self.hidden_dim = hidden_dim
        super(Attention, self).__init__(**kwargs)

    @property
    def inputs_info(self):
        return {'state':2, 'attended':3}
        
    def build(self):
        input_dim = self.inputs['state'].input_shape[1]
        sequence_dim = self.inputs['attended'].input_shape[2]

        self.W_hid = self.init((input_dim + sequence_dim, self.hidden_dim))
        self.W_sft = self.init((self.hidden_dim, 1))
        self.b_hid = self.init((self.hidden_dim, ))
        self.b_sft = self.init((1, ))

        self.params = [self.W_hid, self.W_sft, self.b_hid, self.b_sft]

    @property
    def output_shape(self):
        return (None, 1)

    def get_output(self, train=False):
        ins = self.get_input(train)
        state, attended = ins['state'], ins['attended']
        
        state = state.dimshuffle(0,'x',1)
        repeated_state = T.repeat(state, attended.shape[1], axis=1)
        
        X = T.concatenate((repeated_state, attended), 2)
        
        h = self.activation(T.dot(X, self.W_hid) + self.b_hid)
        pre_softmax = self.activation(T.dot(h, self.W_sft) + self.b_sft)                
        pre_softmax = pre_softmax.reshape((pre_softmax.shape[0], pre_softmax.shape[1]))

        return softmax(pre_softmax)

    def get_config(self):
        return {}

# -------------------------------------------------------------------------------------
    
g = Graph()

g.add_input('X', (100,))
g.add_input('att', (None, 250))

g.add_node(Attention(256), 'attention', input={'state':'X', 'attended':'att'}, create_output=True)

g.compile(Adam(), {'attention':'mse'})
```

ping @EderSantana I'd love to hear Your opinion.

### CODE REVIEW

**Comment by dbonadiman** [comment]
I like the idea seriously, it is more intuitive then the current API even the MultiLayer is a good idea since we are going to have lot of them (At least all the layers i have implemented in my library follows this idea :D )

**Comment by EderSantana** [comment]
nice @elanmart I will check this out soon! You also made me remember its time to resume working on the recurrent container.

**Comment by dbonadiman** [comment]
@EderSantana I thought about it a lot :) for now i'm implementing attentive models with huge repeat vectors, Time distributed merge and Time distributed dense :) Fairly inefficient ^_^

File: keras/layers/containers.py [hunk 1]
```diff
@@ -139,6 +139,22 @@ def __init__(self):
...
+    def _apply(self, ins, train=True):         
+        if self.nb_input == 1:
+            _input = self.inputs[self.inputs_order[0]].get_input
+            self.inputs[self.inputs_order[0]].get_input = lambda _: ins
+            output = self.get_output(train)
+            self.inputs[self.inputs_order[0]].get_input = _input
+        else:
+            _inputs = {}
+            for name, layer in self.inputs.items():
+                _inputs[name] = layer.get_input
```
Comment (EderSantana) [inline keras/layers/containers.py]: double (, layer.get_input()

File: keras/layers/containers.py [hunk 2]
```diff
@@ -239,44 +246,59 @@ def add_input(self, name, input_shape, dtype='float'):
                                   'input_shape': input_shape,
                                   'dtype': dtype})
 
-    def add_node(self, layer, name, input=None, inputs=[],
-                 merge_mode='concat', concat_axis=-1, dot_axes=-1, create_output=False):
+    def add_node(self, layer, name, input=None, merge_mode='concat', concat_axis=-1, create_output=False):
```
Comment (EderSantana) [inline keras/layers/containers.py]: This is a change in API and will break some models. Will need docs and public announcements.

File: keras/layers/core.py [hunk 3]
```diff
@@ -187,6 +199,80 @@ def count_params(self):
         return sum([np.prod(p.shape.eval()) for p in self.params])
 
 
+class MultiInputOutoutLayer(Layer):
```
Comment (EderSantana) [inline keras/layers/core.py]: MultiInputOutputLayer
MIMOLayer is also understandable

File: keras/layers/core.py [hunk 4]
```diff
@@ -187,6 +199,80 @@ def count_params(self):
...
+
+    def fork_inputs(self):
+        self.inputs={}
+        for name, input_dim in self.inputs_info.items():
+            layer = Layer()
+            if input_dim is not None: 
+                layer.input_ndim = input_dim                
+            self.inputs[name] = layer
+            
+    def _filter(self, name):
```
Comment (EderSantana) [inline keras/layers/core.py]: maybe _filter_output for clarity?

File: keras/layers/core.py [hunk 5]
```diff
@@ -1005,7 +1091,7 @@ def __init__(self, function, output_shape=None, **kwargs):
 
     @property
     def output_shape(self):
-        if self._output_shape is None:
+        if self._ouput_shape is None:
```
Comment (EderSantana) [inline keras/layers/core.py]: typo

File: keras/layers/core.py [hunk 6]
```diff
@@ -986,8 +1072,8 @@ class Lambda(Layer):
     output_shape - Expected output shape from function. Could be a tuple or a function of the shape of the input
     """
 
-    def __init__(self, function, output_shape=None, **kwargs):
-        super(Lambda, self).__init__(**kwargs)
+    def __init__(self, function, output_shape=None):
```
Comment (EderSantana) [inline keras/layers/core.py]: why take the kwargs out?

File: keras/layers/containers.py [hunk 7]
```diff
@@ -139,6 +139,22 @@ def __init__(self):
         self.output_config = []  # dicts
         self.node_config = []  # dicts
 
+    def _apply(self, ins, train=True):
```
Comment (EderSantana) [inline keras/layers/containers.py]: Suggestions (that need to be discussed):
* Could we get something like that in a `layer`? Container is a layer, could it inherit from there instead?
* Instead of `_apply`, this could be `__call__`. So we could use the container simply as `model(ins)` instead of `model._apply(ins)`. This is how it is done in TensorFlow for instance.

File: keras/layers/containers.py [hunk 8]
```diff
@@ -179,19 +195,10 @@ def updates(self):
...
-            self.inputs[self.input_order[0]].set_previous(layer)
-        else:
-            if not connection_map:
-                raise Exception('Cannot attach multi-input layer: no connection_map provided.')
-            for k, v in connection_map.items():
-                if k in self.inputs and v in layer.outputs:
-                    self.inputs[k].set_previous(layer.outputs[v])
-                else:
-                    raise Exception('Invalid connection map.')
+    def set_previous(self, layer):
```
Comment (EderSantana) [inline keras/layers/containers.py]: This removes the `connection_map` wasn't it being used somewhere else already? Can you keep it for backward compatibility?

**Comment by EderSantana** [comment]
@elanmart I didn't understand (but that is my fault) how a `MultiInputOutputLayer` connects with a regular layer. If the MIMO layer has only one output, does it work seamlessly? Also, can the outputs of a MIMO be split in a `Graph` model? For example, can two different conventional `Layers` each one get an input from `MIMOLayer`?

TODOs:
* Mostly docs and usage examples.
* Backwards compatibility and deprecation notice.
* Get tests passing.

**Comment by elanmart** [comment]
@EderSantana Thanks for comments!

> If the MIMO layer has only one output, does it work seamlessly?

Yes, you can write:
```python
g = Graph()

g.add_input('X', (100,))
g.add_input('att', (None, 250))

g.add_node(Attention(256), 'attention', input={'state':'X', 'attended':'att'}, create_output=False)
g.add_node(Dense(128), 'd', input='attention', create_output=True)

g.compile(Adam(), {'d':'mse'})
```

> For example, can two different conventional Layers each one get an input from MIMOLayer?

Well, without this the `MIMOLayer` would be useless! 

The way it is done now is what I found most convienient:
given a `MIMOLayer` with name `name` and outputs `out_1`, `out_2`, inside graph two nodes will be created: `name_out_1`, and `name_out_2`. Now each of them is perfectly valid node, so you can plug them into whatever new `Layer` you want.

**Comment by EderSantana** [comment]
got it, thanks for the clarifications!

File: keras/layers/containers.py (hunk 8 above)
Comment (fchollet) [inline keras/layers/containers.py]: Yes, unclear to me why the `connection_map` scheme has to be removed?

File: keras/layers/containers.py [hunk 9]
```diff
@@ -239,44 +246,59 @@ def add_input(self, name, input_shape, dtype='float'):
...
-            for n in inputs:
+            for n in input:
                 if n in self.nodes:
                     to_merge.append(self.nodes[n])
                 elif n in self.inputs:
                     to_merge.append(self.inputs[n])
                 else:
-                    raise Exception('Unknown identifier: ' + n)
-            merge = Merge(to_merge, mode=merge_mode, concat_axis=concat_axis, dot_axes=dot_axes)
+                    raise Exception('Unknown identifier: ' + n)
```
Comment (fchollet) [inline keras/layers/containers.py]: Unnecessary whitespace. Please install a PEP8 linter to spot these issues during development.

File: keras/layers/containers.py [hunk 10]
```diff
@@ -239,44 +246,59 @@ def add_input(self, name, input_shape, dtype='float'):
...
+            for n in input:
                 if n in self.nodes:
                     to_merge.append(self.nodes[n])
                 elif n in self.inputs:
                     to_merge.append(self.inputs[n])
                 else:
-                    raise Exception('Unknown identifier: ' + n)
-            merge = Merge(to_merge, mode=merge_mode, concat_axis=concat_axis, dot_axes=dot_axes)
+                    raise Exception('Unknown identifier: ' + n)                    
+            merge = Merge(to_merge, mode=merge_mode, concat_axis=concat_axis)
```
Comment (fchollet) [inline keras/layers/containers.py]: Why did the dot mode disappear?

**Comment by fchollet** [comment]
> only one input argument

Agreed, that's better. 

> modify set_previous and allow more flexible connectivity using multi-io layers

Can you expand on that?

File: keras/layers/containers.py (hunk 8 above)
Comment (elanmart) [inline keras/layers/containers.py]: Wouldn't this allow us to achieve the same result in two different ways:

```python
graph_A = ... # 2 outputs
graph_B = ... # 2 inputs
model = ... # main graph

model.add_node(graph_A, name='A', ...)
model.add_node(graph_B, input='A', connection_map={
										'input_1': 'output_1', 
										'input_2':'output_2')
# Same as:
model.add_node(graph_B, input={'input_1': 'A_output_1', 'input_2', 'A_output_2'})
```

**Comment by elanmart** [comment]
@fchollet thanks for comments, I'm gonna fix all those small issues ASAP if the main idea gets approved.


>>modify set_previous and allow more flexible connectivity using multi-io layers

>Can you expand on that?

As in [#620(comment)](https://github.com/fchollet/keras/issues/620#issuecomment-144483091)
```python
model.add_node(layer, input={'my_layer_input_1': 'node_name_1', 'my_layer_input_2': 'node_name_2'})
```
only that `node_name_1` is now any `Layer` or any `output` of `MIMOLayer / Graph`. So that you can have a `graph` inside your model, and then feed one of its `outputs` to one `Layer`, and a different `output` to a different `Layer`.

**Comment by fchollet** [comment]
I believe it's a good idea. Since such an interface is assumption-heavy, making the UX work smoothly will require very thorough assumption checking and helpful error messages.

**Comment by elanmart** [comment]
@fchollet what about the `MIMOLayer` itself?

**Comment by farizrahman4u** [comment]
@EderSantana @elanmart Whats the recurrent container?

**Comment by elanmart** [comment]
See #620 

@EderSantana perhaps we could team-up on this one too, I've got some ideas that I think can be quite usefull.

**Comment by EderSantana** [comment]
Oh yeah @farizrahman4u and @elanmart lets do this!!! My sample code worked on the pre-shape inference Keras. But now with the new modifications and so many people familiar with Keras source, it can be much easier to do it. I could make a PR with the old code and update it from there or we could start a new one from scratch. At least we have an API idea now, the implementation will just flow... We have a chance to build the best and most intuitive way to design RNNs. Totally abstracting for-loops whatsoever. Let us continue this conversation on #620 So we don't @elanmart 's PR.

**Comment by bobchennan** [comment]
Any progress on this issue?

**Comment by fchollet** [comment]
Closing outdated PRs following the release of Keras 1.0.
